### PR TITLE
feat(pouchdb): initial adapter

### PR DESCRIPTION
**Pull request by jpbourgeon**
## Reasoning 💡

> A PouchDB adapter for nextauthjs

Depending on your architecture you can :

1. HTTP strategy (serverless or server setup) : use PouchDB's http adapter to reach any database compliant with the CouchDB protocol (CouchDB, Cloudant, ...)
1. Classic strategy (server setup) : use any other PouchDB compatible adapter (in particular leveldb, ...)
1. Memory-first caching strategy (server setup) : if you need to boost your authentication layer performance. Use PouchDB's powerful sync features together with its memory-adapter and any other persistent adapter to build a memory-first caching strategy.

## Checklist 🧢

- [x] Documentation
- [x] Tests
- [ ] Ready to be merged

## Affected issues 🎟

None

### CODE REVIEW

**Comment by jpbourgeon** [comment]
I'm a total noob with Typescript. I did my best to comply with typing the adapter but it's probably weak.
Any advice or direct edit on that matter (among others) is welcome !

**Comment by balazsorban44** [comment]
Thank you! I am looking forward to reviewing this! We are currently figuring out some stuff about the adapters, so I'll leave this open for now, but at a first glance, it looks very good!

**Comment by jpbourgeon** [comment]
Hi
Do you have an ETA of releasing new adapters structure ? I would then fix this adapter to match the new format.
Best!

**Comment by balazsorban44** [comment]
totally, you can take this adapter as a relatively good example: https://www.github.com/nextauthjs/adapters/tree/canary/packages%2Fprisma%2Fsrc%2Findex.ts

**Comment by ndom91** [comment]
@jpbourgeon we've got the new adapter structure and basic tests up and running.

Adapter TS definition: https://github.com/nextauthjs/next-auth/blob/main/types/adapters.d.ts
`basic-tests.ts` for all adapters: https://github.com/nextauthjs/adapters/blob/canary/basic-tests.ts

Looking forward to this!

**Comment by jpbourgeon** [comment]
Hi

I updated the pouchdb adapter PR with the following:

- typescript implementation following the new Adapter TS definition.
- fully tested against the basic-tests testsuite. I also provided temporary tests until the skipped deleteUser and unlinkAccount basic tests are provided.
- I took the liberty to remove the assertions on createdAt and updatedAt properties in the linkAccount basic test, since the tests of other methods don't assert on createdAt and updatedAt, and since some other database engines (including pouchdb) don't manage these properties automatically. Should you say otherwise, I would revert this change and manage the createdAt and updatedAt properties on all the adapter's objects manually.

I look forward to your feedback!

**Comment by balazsorban44** [comment]
Thank you! Will follow this up when I have the time :slightly_smiling_face:

File: packages/pouchdb/README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# PouchDB Adapter
```
Comment (ndom91) [inline packages/pouchdb/README.md:1]: Dig all the work you put into updating the READMEs, package.json, etc. as well!

Small nit though. Lets make this header consistent with the other ones. Something like:

```markdown
<p align="center">
   <br/>
   <a href="https://next-auth.js.org" target="_blank">
   <img height="64px" src="https://next-auth.js.org/img/logo/logo-sm.png" /></a>&nbsp;&nbsp;&nbsp;&nbsp;<img height="64px" src="https://raw.githubusercontent.com/nextauthjs/adapters/canary/packages/pouchdb/logo.svg" />
   <h3 align="center"><b>PouchDB Adapter</b> - NextAuth.js</h3>
   <p align="center">
   Open Source. Full Stack. Own Your Data.
   </p>
   <p align="center" style="align: center;">
      <img src="https://github.com/nextauthjs/adapters/actions/workflows/canary.yml/badge.svg" alt="Build Test" />
      <img src="https://img.shields.io/bundlephobia/minzip/@next-auth/pouchdb-adapter" alt="Bundle Size"/>
      <img src="https://img.shields.io/npm/v/@next-auth/pouchdb-adapter" alt="@next-auth/pouchdb-adapter Version" />
   </p>
</p>

## Overview
```

See [fauna/readme.md](https://raw.githubusercontent.com/nextauthjs/adapters/canary/packages/fauna/README.md) for example.

Comment (jpbourgeon) [reply packages/pouchdb/README.md]: I updated the doc according to the format of the other adapters docs.

**Comment by ndom91** [comment]
Great PR :clap: :clap:

**Comment by jpbourgeon** [comment]
Hi
How can I help you drive that nail ?
I may write a new adapter for the leveldow/leveldb family but I would like to finish this one beforehand !

**Comment by balazsorban44** [comment]
There is a yarn.lock merge conflict, could you have a look at that? I hope I'll have time to give you feedback on this PR this weekend maybe.

**Comment by jpbourgeon** [comment]
> There is a yarn.lock merge conflict, could you have a look at that? I hope I'll have time to give you feedback on this PR this weekend maybe.

I updated the PR 😸

**Comment by ndom91** [comment]
Thanks for all your hardwork @jpbourgeon! I'm going to merge this now - the pouchdb tests all seem to be passing!

**Pull request merged**